lua/ttt2/libraries/orm.lua

```diff
@@ -0,0 +1,219 @@
+---
+-- Object Relational Model
+-- @author Histalek
+
+if SERVER then
+	AddCSLuaFile()
+end
+
+local sql = sql
+
+orm = orm or {}
+
+local ORMMODEL = {}
+
+---
+-- Returns an object relational model according to the specified databasetable. Does nothing if no databasetable with the given name exists.
+-- @param string tableName The name of the table in the database to create a model for.
+-- @param[opt] boolean force If set to `true` the function will not return a cached version of the model.
+-- @return ormmodel The model of the database table.
+-- @realm shared
+function orm.Make(tableName, force)
+	if IsValid(orm[tableName]) and not force then
+		return orm[tableName]
+	end
+
+	if not sql.TableExists(tableName) then return end
+
+	local model = {}
+	local primaryKey = sql.GetPrimaryKey(tableName)
+	local dataStructure = sql.GetTableColumns(tableName)
+
+	model._tableName = tableName
+	model._primaryKey = primaryKey
+	model._dataStructure = dataStructure
+	model.All = ORMMODEL.All
+	model.Find = ORMMODEL.Find
+	model.New = ORMMODEL.New
+
+	-- DO NOT setup delete/save functions if no primarykey is found.
+	-- In those cases the 'rowid' column would function as the primarykey, but as the rowid could change anytime (https://www.sqlite.org/rowidtable.html) data could be deleted unintentionally.
+	-- Most likely rowids wont change in gmod as there is no vacuum operation but just to be safe we will not allow to use such tables. ref: https://wiki.facepunch.com/gmod/sql
+
+	if not primaryKey then
+		return model
+	end
+
+	model.Delete = ORMMODEL.Delete
```

Comment:
Should these functions really be callable on the generic model without a primary key? (Or should this be implemented to delete all content of a table?)

Reply:
i'm not sure how i would prevent this though. Furthermore these functions called on the generic model will only create an sql error and not actually change any content but yeah it doesn't really make sense to have them callable there

Reply:
PTAL i split the class into two to make them more distinct which has the sideeffect of solving this problem